common/cuda_hip/reorder/rcm_kernels.hpp.inc

```diff
@@ -0,0 +1,608 @@
+        }
+    }
+
+    const IndexType* candidate_node;
+    const IndexType* level;
+    IndexType* best_node;
+    IndexType* best_level;
+    bool* improved;
+};
+
+
+template <typename IndexType>
+void find_pseudo_peripheral_nodes(std::shared_ptr<const DefaultExecutor> exec,
+                                  const IndexType num_rows,
+                                  const IndexType* const row_ptrs,
+                                  const IndexType* const col_idxs,
+                                  const IndexType* const node_degrees,
+                                  const components_data<IndexType>& components,
+                                  ubfs_levels<IndexType>& levels)
+{
+    const auto policy = thrust_policy(exec);
+    const auto num_components = components.get_num_components();
+    array<IndexType> candidate_node_array{exec, num_components};
+    array<IndexType> best_level_array{exec, num_components};
+    array<bool> improved{exec, 1};
+    const auto candidate_nodes = candidate_node_array.get_data();
+    const auto best_levels = best_level_array.get_data();
+    const auto level_nodes = levels.nodes.get_data();
+    const auto node_levels = levels.node_level.get_const_data();
+    const auto component_nodes = components.nodes.get_const_data();
+    const auto sorted_component_ids = components.sorted_ids.get_const_data();
+    const auto reduction = node_max_level_min_degree_reduction<IndexType>{
+        num_rows, node_degrees, node_levels};
+    const auto discard_it = thrust::discard_iterator<IndexType>{};
+    const auto eq_op = thrust::equal_to<IndexType>{};
+    const auto counting_it = thrust::make_counting_iterator(IndexType{});
+    const auto compare_fn = node_compare_functor<IndexType>{
+        candidate_nodes, node_levels, level_nodes, best_levels,
+        improved.get_data()};
+    // initialize best_levels and levels to the initial nodes at level 0
+    thrust::fill_n(policy, best_levels, num_components, IndexType{});
+    do {
+        ubfs(exec, num_rows, row_ptrs, col_idxs, levels);
+        // write a last-level node of min degree to candidate_nodes for each
+        // component
+        thrust::reduce_by_key(policy, sorted_component_ids,
+                              sorted_component_ids + num_rows, component_nodes,
+                              discard_it, candidate_nodes, eq_op, reduction);
+        set_element(improved, 0, false);
+        thrust::for_each_n(policy, counting_it, num_components, compare_fn);
+    } while (get_element(improved, 0));
+    // the best nodes stay on the 0th level
+}
+
+
+template <typename IndexType>
+__global__ __launch_bounds__(default_block_size) void ubfs_min_neighbor_kernel(
+    const IndexType* __restrict__ row_ptrs,
+    const IndexType* __restrict__ col_idxs, const IndexType level_begin,
+    const IndexType level_end, const IndexType* __restrict__ level_nodes,
+    const IndexType* __restrict__ inv_permutation,
```

Comment:
This is not used. So the `construct_inverse_permutation` factory parameter is effectively ignored.